src/LinAlg/hiopMatrixSparseCSRSeq.hpp

```diff
@@ -0,0 +1,598 @@
+   */
+  void form_diag_from_symbolic(const hiopVector& D);
+  
+  /**
+   * Sets the diagonal entries of `this` equal to entries of D
+   * 
+   * @pre Length of `D` should be equal to size(s) of `this`
+   * 
+   * @pre `this` should be a diagonal matrix (in CSR format) with storage for
+   * all the diagonal entries, which can be ensured by calling the sister method
+   * `form_diag_from_symbolic`
+   */
+  void form_diag_from_numeric(const hiopVector& D);
+  
+  /**
+   * Allocates and returns CSR matrix `M` capable of holding M = X+Y, where X is 
+   * the calling matrix class (`this`) and Y is the argument passed to the method.
+   */
+  hiopMatrixSparseCSR* add_matrix_alloc(const hiopMatrixSparseCSR& Y) const;
+
+  /**
+   * Computes sparsity pattern of M = X+Y (i.e., populates the row pointers and 
+   * column indexes arrays) of `M`.
+   * 
+   * @pre `this` and `Y` should hold matrices of identical dimensions.
+   *
+   */
+
+  //// note cusparseXcsrgeam2
+  void add_matrix_symbolic(hiopMatrixSparseCSR& M, const hiopMatrixSparseCSR& Y) const;
+
+  /**
+   * Performs matrix addition M = gamma*M + alpha*X + beta*Y numerically, where
+   * X is `this` and gamma, alpha, and beta are scalars.
+   * 
+   * @pre `M`, `this` and `Y` should hold matrices of identical dimensions.
+   * 
+   * @pre `M` and `X+Y` should have identical sparsity pattern, namely the 
+   * `add_matrix_symbolic` should have been called previously.
+   *
+   */
+  void add_matrix_numeric(double gamma,
+                          hiopMatrixSparseCSR& M,
+                          double alpha,
+                          const hiopMatrixSparseCSR& Y,
+                          double beta) const;
+
+  /// @brief Performs a quick check and returns false if the CSR indexes are not ordered
+  bool check_csr_is_ordered();
+  /////////////////////////////////////////////////////////////////////
+  // end of new CSR-specific methods
+  /////////////////////////////////////////////////////////////////////
+
+private:
+  void alloc();
+  void dealloc();
+protected:
+
+  //// inherits nrows_, ncols_, and nnz_ from  hiopSparseMatrix
+  
+  /// Row pointers (starting indexes) in the column and values arrays
+  index_type* irowptr_;
+
+  /// Column indexes
+  index_type* jcolind_;
+
+  /// Nonzero values
+  double* values_;
+
+  /// Working buffer in the size of columns, allocated on demand and reused by some methods
+  double* buf_col_;
+
+  /**
+   * Storage for the row starts used by `form_transpose_from_xxx` methods (allocated on 
+   * demand, only the above mentioned methods are called)
+   */
+  index_type* row_starts_;
+  
+private:
+  hiopMatrixSparseCSRSeq(const hiopMatrixSparseCSRSeq&)
+    : hiopMatrixSparseCSR(0, 0, 0),
+      irowptr_(nullptr),
+      jcolind_(nullptr),
+      values_(nullptr),
+      buf_col_(nullptr),
+      row_starts_(nullptr)
+  {
+    assert(false && "copy constructor disabled");
+  }
```

Comment:
```suggestion
  hiopMatrixSparseCSRSeq(const hiopMatrixSparseCSRSeq&) = delete;
```

Reply:
Why is copy constructor disabled? You could make it private and then use it to implement `clone` and `dep_copy` methods, which can be handy.